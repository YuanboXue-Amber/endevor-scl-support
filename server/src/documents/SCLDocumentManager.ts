import {
    TextDocument,
    TextDocumentItem,
    VersionedTextDocumentIdentifier,
    TextDocumentContentChangeEvent,
    TextDocumentPositionParams,
    CompletionItem,
    TextDocumentIdentifier,
    TextEdit,
    CompletionItemKind,
    Position, Range,
    Command,
    CodeLens,
    ExecuteCommandParams} from "vscode-languageserver";
import { SCLDocument } from './SCLDocument';
import { isNull, isNullOrUndefined } from "util";
import { isUnaryLike } from "@babel/types";
import { commands, executeSubmitSCL } from "../ExecuteCommandProvider";

export interface IDocumentSettings {
    maxNumberOfProblems: number;
    isREST: boolean;
}

interface ICapabilities {
    hasDiagnosticRelatedInformationCapability: boolean;
}

export const actionCompletion = [ "SET", "ADD", "UPDATE", "DELETE", "GENERATE",
"MOVE", "RETRIEVE", "SIGNIN", "TRANSFER", "APPROVE", "DENY", "BACKIN",
"BACKOUT", "CAST", "DEFINE", "EXECUTE", "RESET", "COMMIT", "LIST"];

/**
 * Manage all the opened documents in the client
 *
 * @export
 * @class SCLDocumentManager
 */
export class SCLDocumentManager {
    static config: IDocumentSettings = { maxNumberOfProblems: 1000, isREST: false };
    static capabilities: ICapabilities = { hasDiagnosticRelatedInformationCapability: false };

    static numberOfProblems = 0;

    // a map of document uri to document details
    documents: Map<string, SCLDocument> = new Map();

    constructor() {
        this.documents = new Map();
    }

    // NEVER USED
    openDocument(textDocumentItem: TextDocumentItem): SCLDocument {
        const textDocument: TextDocument = TextDocument.create(
            textDocumentItem.uri, textDocumentItem.languageId, textDocumentItem.version, textDocumentItem.text);
        const document: SCLDocument = new SCLDocument(textDocument);
        this.documents.set(textDocument.uri, document);

        // no need to refresh the diagnose number, since it is dealt with in SCLDocument.pushDiagnostic
        return document;
    }

    openOrChangeDocument(textDocument: TextDocument): SCLDocument {
        let document = this.documents.get(textDocument.uri);
        if (!document) {
            document = new SCLDocument(textDocument);
            this.documents.set(textDocument.uri, document);
            return document;
        }

        // reset the diagnose number
        let decreaseNum = 0;
        document.statements.forEach((statement) => {
            decreaseNum += statement.diagnostics.length;
        });
        SCLDocumentManager.numberOfProblems = SCLDocumentManager.numberOfProblems - decreaseNum;
        if (SCLDocumentManager.numberOfProblems < 0) {
            SCLDocumentManager.numberOfProblems = 0;
        }

        document.textDocument = textDocument;
        document.fullUpdate();
        this.documents.set(textDocument.uri, document);

        return document;
    }

    closeDocument(textDocumentUri: string): void {
        const document = this.documents.get(textDocumentUri);
        if (isNullOrUndefined(document)) {
            return;
        }

        // refresh the diagnose number
        let decreaseNum = 0;
        document.statements.forEach((statement) => {
            decreaseNum += statement.diagnostics.length;
        });
        SCLDocumentManager.numberOfProblems = SCLDocumentManager.numberOfProblems - decreaseNum;
        if (SCLDocumentManager.numberOfProblems < 0) {
            SCLDocumentManager.numberOfProblems = 0;
        }

        this.documents.delete(textDocumentUri);
    }

    getCompletionBySyntax(textDocumentPosition: TextDocumentPositionParams): CompletionItem[] {
        const document = this.documents.get(textDocumentPosition.textDocument.uri);
        if (!document) {
            throw new Error(`Cannot provide completion on unopened document: ${textDocumentPosition.textDocument.uri}`);
        }

        const tobeCompelteIndex = document.textDocument.offsetAt(textDocumentPosition.position);
        let completionItems: CompletionItem[] = [];

        for (const statement of document.statements) {
            if (tobeCompelteIndex >= statement.starti && tobeCompelteIndex <= statement.endi) {
                let i = -1;
                while (i < statement.tokens.length - 1) {
                    i ++;
                    const starti = statement.tokens[i].starti + statement.tokens[i].value.length;
                    const endi = (i+1) < statement.tokens.length ? statement.tokens[i+1].starti : statement.endi;

                    if (tobeCompelteIndex >= starti && tobeCompelteIndex <= endi) {
                        // to complete between tokens[i] and tokens[i+1] in scl
                        const values = statement.tokens[i].completionItems;
                        if (!isNullOrUndefined(values)) {
                            if (statement.tokens[i].value !== ".") {
                                completionItems = completionItems.concat(values);
                            } else {
                                for (const each of actionCompletion) {
                                    completionItems.push({
                                        label: each + " ",
                                        kind: CompletionItemKind.Function,
                                        documentation: "Endevor SCL keyword"
                                    });
                                }
                            }
                        }
                        return completionItems;
                    }
                }
            }
        }

        // to complete at the end of scl
        const statement = document.statements[document.statements.length-1];
        let values;
        if (!isNullOrUndefined(statement))
            values = statement.tokens[statement.tokens.length-1].completionItems;
        if (!isNullOrUndefined(values) && statement.tokens[statement.tokens.length-1].value !== ".") {
            completionItems = completionItems.concat(values);
        } else {
            // maybe a new scl is started
            for (const each of actionCompletion) {
                completionItems.push({
                    label: each + " ",
                    kind: CompletionItemKind.Function,
                    documentation: "Endevor SCL keyword"
                });
            }
        }
        return completionItems;
    }

    async formatDocument(textDocument: TextDocumentIdentifier): Promise<TextEdit[]> {
        const document = this.documents.get(textDocument.uri);
        if (!document) {
            throw new Error(`Cannot provide formatting on unopened document: ${textDocument.uri}`);
        }

        return document.formatDocument();
    }

    async computeCodeLenses(textDocument: TextDocumentIdentifier): Promise<CodeLens[]> {
        const document = this.documents.get(textDocument.uri);
        if (!document) {
            throw new Error(`Cannot provide codeLens on unopened document: ${textDocument.uri}`);
        }

        let codeLens: CodeLens[] = [];
        for (const statement of document.statements) {
            const range = {
                start: document.textDocument.positionAt(statement.starti),
                end: document.textDocument.positionAt(statement.starti+1),
            };
            codeLens = codeLens.concat(
                commands.map(command => ({
                    range,
                    command: Command.create(command.title, command.command, textDocument, range)
                }))
            );
        }
        return codeLens;
    }

    async executeCodeLens(params: ExecuteCommandParams) {
        const textDocument: TextDocumentIdentifier = (params.arguments as any[])[0];
        const range: Range = (params.arguments as any[])[1];
        const document = this.documents.get(textDocument.uri);
        if (!document) {
            throw new Error(`Cannot provide codeLens on unopened document: ${textDocument.uri}`);
        }
        return executeSubmitSCL(document, document.textDocument.offsetAt(range.start));
    }

}