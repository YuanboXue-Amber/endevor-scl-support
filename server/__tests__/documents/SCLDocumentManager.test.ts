import { SCLDocumentManager } from '../../src/documents/SCLDocumentManager';
import { TextDocumentItem, VersionedTextDocumentIdentifier, TextDocumentContentChangeEvent } from 'vscode-languageserver';
import { SCLDocument } from '../../src/documents/SCLDocument';

describe("Test", () => {

    it("test1: open document", async () => {

        const openedDocument: TextDocumentItem = {
            uri: "mocked/uri",
            languageId: "mocklanguageId",
            version: 1,
            text: "SET action add . \n SET ACTION RESTORE. \n Set INvalid. "
        };
        const manager = new SCLDocumentManager();
        const document: SCLDocument = manager.openDocument(openedDocument);
        expect(document.textDocument.getText()).toMatchSnapshot();
        for (const statement of document.statements) {
            expect(statement.tokens).toMatchSnapshot();
            expect(statement.starti).toMatchSnapshot();
            expect(statement.endi).toMatchSnapshot();
            expect(statement.diagnostics).toMatchSnapshot();
        }
    });

    it("test2: update document", async () => {

        const updatedDocumentId: VersionedTextDocumentIdentifier = {
            uri: "mocked/uri",
            version: 1
        };

        const contentChanges: TextDocumentContentChangeEvent[] = [
            {
                // substitude at the end
                range: {
                    start: {
                        line: 2,
                        character: 5
                    },
                    end: {
                        line: 2,
                        character: 12
                    }
                },
                rangeLength: 1,
                text: "ACTION TRANSFER "
            },
            {
                // change the line ending after "SET ACTION RESTORE. "
                range: {
                    start: {
                        line: 1,
                        character: 21
                    },
                    end: {
                        line: 1,
                        character: 22
                    }
                },
                rangeLength: 1,
                text: "."
            },
            {
                // insert word SET at the beginning
                range: {
                    start: {
                        line: 0,
                        character: 0
                    },
                    end: {
                        line: 0,
                        character: 0
                    }
                },
                rangeLength: 0,
                text: "SET"
            }
        ];

        const openedDocument: TextDocumentItem = {
            uri: "mocked/uri",
            languageId: "mocklanguageId",
            version: 1,
            text: "SET action add . \n SET ACTION RESTORE. \n Set INvalid. "
        };
        const manager = new SCLDocumentManager();
        manager.openDocument(openedDocument);
        const document: SCLDocument = manager.updateDocument(updatedDocumentId, contentChanges);
        expect(document.textDocument.getText()).toMatchSnapshot();
        for (const statement of document.statements) {
            expect(statement.tokens).toMatchSnapshot();
            expect(statement.starti).toMatchSnapshot();
            expect(statement.endi).toMatchSnapshot();
            expect(statement.diagnostics).toMatchSnapshot();
        }
    });

    it("test3: update document", async () => {

        const updatedDocumentId: VersionedTextDocumentIdentifier = {
            uri: "mocked/uri",
            version: 1
        };

        const contentChanges: TextDocumentContentChangeEvent[] = [
            {
                // insert at the end
                range: {
                    start: {
                        line: 4,
                        character: 1
                    },
                    end: {
                        line: 4,
                        character: 1
                    }
                },
                rangeLength: 0,
                text: " SET ACTION transfer ."
            },
            {
                // remove the line ending after "SET ACTION RESTORE. "
                range: {
                    start: {
                        line: 2,
                        character: 17
                    },
                    end: {
                        line: 2,
                        character: 18
                    }
                },
                rangeLength: 1,
                text: ""
            },
            {
                // remove SET.
                range: {
                    start: {
                        line: 1,
                        character: 6
                    },
                    end: {
                        line: 1,
                        character: 10
                    }
                },
                rangeLength: 4,
                text: ""
            },
            {
                // substitude action \nadd to action DELETE
                range: {
                    start: {
                        line: 0,
                        character: 4
                    },
                    end: {
                        line: 1,
                        character: 3
                    }
                },
                rangeLength: 11,
                text: "action DELETE"
            }
        ];

        const openedDocument: TextDocumentItem = {
            uri: "mocked/uri",
            languageId: "mocklanguageId",
            version: 1,
            text: "SET action \nadd . SET. SET \n ACTION RESTORE. \n Set ACTION invalid\n. "
        };
        const manager = new SCLDocumentManager();
        manager.openDocument(openedDocument);
        const document: SCLDocument = manager.updateDocument(updatedDocumentId, contentChanges);
        expect(document.textDocument.getText()).toMatchSnapshot();
        for (const statement of document.statements) {
            expect(statement.tokens).toMatchSnapshot();
            expect(statement.starti).toMatchSnapshot();
            expect(statement.endi).toMatchSnapshot();
            expect(statement.diagnostics).toMatchSnapshot();
        }
    });

    it("test4: close document", async () => {

        const openedDocument: TextDocumentItem = {
            uri: "mocked/uri",
            languageId: "mocklanguageId",
            version: 1,
            text: "SET action \nadd . SET. SET \n ACTION RESTORE. \n Set ACTION invalid\n. "
        };
        const manager = new SCLDocumentManager();
        const document = manager.openDocument(openedDocument);
        expect(document.textDocument.getText()).toMatchSnapshot();

        manager.closeDocument(openedDocument.uri);
        expect(manager.documents.size).toBe(0);
    });

    it("test5: update document", async () => {

        const updatedDocumentId: VersionedTextDocumentIdentifier = {
            uri: "mocked/uri",
            version: 1
        };

        const contentChanges: TextDocumentContentChangeEvent[] = [
            {
                // insert at the end
                range: {
                    start: {
                        line: 0,
                        character: 21
                    },
                    end: {
                        line: 0,
                        character: 21
                    }
                },
                rangeLength: 0,
                text: "ACTION TRANSFER ."
            }
        ];

        const openedDocument: TextDocumentItem = {
            uri: "mocked/uri",
            languageId: "mocklanguageId",
            version: 1,
            text: "SET action add . SET "
        };
        const manager = new SCLDocumentManager();
        manager.openDocument(openedDocument);
        const document: SCLDocument = manager.updateDocument(updatedDocumentId, contentChanges);
        expect(document.textDocument.getText()).toMatchSnapshot();
        for (const statement of document.statements) {
            expect(statement.tokens).toMatchSnapshot();
            expect(statement.starti).toMatchSnapshot();
            expect(statement.endi).toMatchSnapshot();
            expect(statement.diagnostics).toMatchSnapshot();
        }
    });

});