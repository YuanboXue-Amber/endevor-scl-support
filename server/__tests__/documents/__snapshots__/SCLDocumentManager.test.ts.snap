// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test test1: open document 1`] = `
"SET action add . 
 SET ACTION RESTORE. 
 Set INvalid. "
`;

exports[`Test test1: open document 2`] = `
Array [
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 0,
    "value": "SET",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 4,
    "value": "action",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 11,
    "value": "add",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": true,
    "is_op_char": false,
    "is_word": false,
    "starti": 15,
    "value": ".",
  },
]
`;

exports[`Test test1: open document 3`] = `0`;

exports[`Test test1: open document 4`] = `16`;

exports[`Test test1: open document 5`] = `
Array [
  Object {
    "diagnostic": Object {
      "message": "Keyword should be uppercased",
      "range": Object {
        "end": Object {
          "character": 10,
          "line": 0,
        },
        "start": Object {
          "character": 4,
          "line": 0,
        },
      },
      "severity": 2,
      "source": "Endevor SCL extension",
    },
    "endi": 10,
    "starti": 4,
  },
  Object {
    "diagnostic": Object {
      "message": "Keyword should be uppercased",
      "range": Object {
        "end": Object {
          "character": 14,
          "line": 0,
        },
        "start": Object {
          "character": 11,
          "line": 0,
        },
      },
      "severity": 2,
      "source": "Endevor SCL extension",
    },
    "endi": 14,
    "starti": 11,
  },
]
`;

exports[`Test test1: open document 6`] = `
Array [
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 19,
    "value": "SET",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 23,
    "value": "ACTION",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 30,
    "value": "RESTORE",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": true,
    "is_op_char": true,
    "is_word": false,
    "starti": 37,
    "value": ".",
  },
]
`;

exports[`Test test1: open document 7`] = `19`;

exports[`Test test1: open document 8`] = `38`;

exports[`Test test1: open document 9`] = `
Array [
  Object {
    "diagnostic": Object {
      "message": "Invalid scl. Expecting a space before end of statement operator \\".\\"",
      "range": Object {
        "end": Object {
          "character": 20,
          "line": 1,
        },
        "start": Object {
          "character": 19,
          "line": 1,
        },
      },
      "severity": 1,
      "source": "Endevor SCL extension",
    },
    "endi": 38,
    "starti": 37,
  },
]
`;

exports[`Test test1: open document 10`] = `
Array [
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 41,
    "value": "Set",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 45,
    "value": "INvalid",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": true,
    "is_op_char": true,
    "is_word": false,
    "starti": 52,
    "value": ".",
  },
]
`;

exports[`Test test1: open document 11`] = `41`;

exports[`Test test1: open document 12`] = `53`;

exports[`Test test1: open document 13`] = `
Array [
  Object {
    "diagnostic": Object {
      "message": "Keyword should be uppercased",
      "range": Object {
        "end": Object {
          "character": 4,
          "line": 2,
        },
        "start": Object {
          "character": 1,
          "line": 2,
        },
      },
      "severity": 2,
      "source": "Endevor SCL extension",
    },
    "endi": 44,
    "starti": 41,
  },
  Object {
    "diagnostic": Object {
      "message": "Invalid keyword \\"INvalid\\" specified after SET",
      "range": Object {
        "end": Object {
          "character": 12,
          "line": 2,
        },
        "start": Object {
          "character": 5,
          "line": 2,
        },
      },
      "severity": 1,
      "source": "Endevor SCL extension",
    },
    "endi": 52,
    "starti": 45,
  },
]
`;

exports[`Test test2: update document 1`] = `
"SETSET action add . 
 SET ACTION RESTORE. . Set ACTION TRANSFER . "
`;

exports[`Test test2: update document 2`] = `
Array [
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 22,
    "value": "SET",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 26,
    "value": "ACTION",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 33,
    "value": "RESTORE",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": true,
    "is_op_char": true,
    "is_word": false,
    "starti": 40,
    "value": ".",
  },
]
`;

exports[`Test test2: update document 3`] = `22`;

exports[`Test test2: update document 4`] = `41`;

exports[`Test test2: update document 5`] = `
Array [
  Object {
    "diagnostic": Object {
      "message": "Invalid scl. Expecting a space before end of statement operator \\".\\"",
      "range": Object {
        "end": Object {
          "character": 20,
          "line": 1,
        },
        "start": Object {
          "character": 19,
          "line": 1,
        },
      },
      "severity": 1,
      "source": "Endevor SCL extension",
    },
    "endi": 41,
    "starti": 40,
  },
]
`;

exports[`Test test2: update document 6`] = `
Array [
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 44,
    "value": "Set",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 48,
    "value": "ACTION",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 55,
    "value": "TRANSFER",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": true,
    "is_op_char": false,
    "is_word": false,
    "starti": 64,
    "value": ".",
  },
]
`;

exports[`Test test2: update document 7`] = `44`;

exports[`Test test2: update document 8`] = `65`;

exports[`Test test2: update document 9`] = `
Array [
  Object {
    "diagnostic": Object {
      "message": "Keyword should be uppercased",
      "range": Object {
        "end": Object {
          "character": 26,
          "line": 1,
        },
        "start": Object {
          "character": 23,
          "line": 1,
        },
      },
      "severity": 2,
      "source": "Endevor SCL extension",
    },
    "endi": 47,
    "starti": 44,
  },
]
`;

exports[`Test test2: update document 10`] = `
Array [
  Object {
    "is_eoInput": false,
    "is_eoStatement": true,
    "is_op_char": false,
    "is_word": false,
    "starti": 42,
    "value": ".",
  },
]
`;

exports[`Test test2: update document 11`] = `42`;

exports[`Test test2: update document 12`] = `43`;

exports[`Test test2: update document 13`] = `Array []`;

exports[`Test test2: update document 14`] = `
Array [
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 0,
    "value": "SETSET",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 7,
    "value": "action",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 14,
    "value": "add",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": true,
    "is_op_char": false,
    "is_word": false,
    "starti": 18,
    "value": ".",
  },
]
`;

exports[`Test test2: update document 15`] = `0`;

exports[`Test test2: update document 16`] = `19`;

exports[`Test test2: update document 17`] = `Array []`;

exports[`Test test3: update document 1`] = `
"SET action DELETE .  SET 
 ACTION RESTORE.  Set ACTION invalid
. SET ACTION transfer . "
`;

exports[`Test test3: update document 2`] = `
Array [
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 21,
    "value": "SET",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 27,
    "value": "ACTION",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 34,
    "value": "RESTORE",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": true,
    "is_op_char": true,
    "is_word": false,
    "starti": 41,
    "value": ".",
  },
]
`;

exports[`Test test3: update document 3`] = `21`;

exports[`Test test3: update document 4`] = `42`;

exports[`Test test3: update document 5`] = `
Array [
  Object {
    "diagnostic": Object {
      "message": "Invalid scl. Expecting a space before end of statement operator \\".\\"",
      "range": Object {
        "end": Object {
          "character": 16,
          "line": 1,
        },
        "start": Object {
          "character": 15,
          "line": 1,
        },
      },
      "severity": 1,
      "source": "Endevor SCL extension",
    },
    "endi": 42,
    "starti": 41,
  },
]
`;

exports[`Test test3: update document 6`] = `
Array [
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 44,
    "value": "Set",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 48,
    "value": "ACTION",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 55,
    "value": "invalid",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": true,
    "is_op_char": false,
    "is_word": false,
    "starti": 63,
    "value": ".",
  },
]
`;

exports[`Test test3: update document 7`] = `44`;

exports[`Test test3: update document 8`] = `64`;

exports[`Test test3: update document 9`] = `
Array [
  Object {
    "diagnostic": Object {
      "message": "Keyword should be uppercased",
      "range": Object {
        "end": Object {
          "character": 21,
          "line": 1,
        },
        "start": Object {
          "character": 18,
          "line": 1,
        },
      },
      "severity": 2,
      "source": "Endevor SCL extension",
    },
    "endi": 47,
    "starti": 44,
  },
  Object {
    "diagnostic": Object {
      "message": "Invalid action \\"invalid\\" specified after SET ACTION",
      "range": Object {
        "end": Object {
          "character": 36,
          "line": 1,
        },
        "start": Object {
          "character": 29,
          "line": 1,
        },
      },
      "severity": 1,
      "source": "Endevor SCL extension",
    },
    "endi": 62,
    "starti": 55,
  },
]
`;

exports[`Test test3: update document 10`] = `
Array [
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 65,
    "value": "SET",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 69,
    "value": "ACTION",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 76,
    "value": "transfer",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": true,
    "is_op_char": false,
    "is_word": false,
    "starti": 85,
    "value": ".",
  },
]
`;

exports[`Test test3: update document 11`] = `65`;

exports[`Test test3: update document 12`] = `86`;

exports[`Test test3: update document 13`] = `
Array [
  Object {
    "diagnostic": Object {
      "message": "Keyword should be uppercased",
      "range": Object {
        "end": Object {
          "character": 21,
          "line": 2,
        },
        "start": Object {
          "character": 13,
          "line": 2,
        },
      },
      "severity": 2,
      "source": "Endevor SCL extension",
    },
    "endi": 84,
    "starti": 76,
  },
]
`;

exports[`Test test3: update document 14`] = `
Array [
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 0,
    "value": "SET",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 4,
    "value": "action",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 11,
    "value": "DELETE",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": true,
    "is_op_char": false,
    "is_word": false,
    "starti": 18,
    "value": ".",
  },
]
`;

exports[`Test test3: update document 15`] = `0`;

exports[`Test test3: update document 16`] = `19`;

exports[`Test test3: update document 17`] = `
Array [
  Object {
    "diagnostic": Object {
      "message": "Keyword should be uppercased",
      "range": Object {
        "end": Object {
          "character": 10,
          "line": 0,
        },
        "start": Object {
          "character": 4,
          "line": 0,
        },
      },
      "severity": 2,
      "source": "Endevor SCL extension",
    },
    "endi": 10,
    "starti": 4,
  },
]
`;

exports[`Test test4: close document 1`] = `
"SET action 
add . SET. SET 
 ACTION RESTORE. 
 Set ACTION invalid
. "
`;

exports[`Test test5: update document 1`] = `"SET action add . SET ACTION TRANSFER ."`;

exports[`Test test5: update document 2`] = `
Array [
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 0,
    "value": "SET",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 4,
    "value": "action",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 11,
    "value": "add",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": true,
    "is_op_char": false,
    "is_word": false,
    "starti": 15,
    "value": ".",
  },
]
`;

exports[`Test test5: update document 3`] = `0`;

exports[`Test test5: update document 4`] = `16`;

exports[`Test test5: update document 5`] = `
Array [
  Object {
    "diagnostic": Object {
      "message": "Keyword should be uppercased",
      "range": Object {
        "end": Object {
          "character": 10,
          "line": 0,
        },
        "start": Object {
          "character": 4,
          "line": 0,
        },
      },
      "severity": 2,
      "source": "Endevor SCL extension",
    },
    "endi": 10,
    "starti": 4,
  },
  Object {
    "diagnostic": Object {
      "message": "Keyword should be uppercased",
      "range": Object {
        "end": Object {
          "character": 14,
          "line": 0,
        },
        "start": Object {
          "character": 11,
          "line": 0,
        },
      },
      "severity": 2,
      "source": "Endevor SCL extension",
    },
    "endi": 14,
    "starti": 11,
  },
]
`;

exports[`Test test5: update document 6`] = `
Array [
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 17,
    "value": "SET",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 21,
    "value": "ACTION",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 28,
    "value": "TRANSFER",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": true,
    "is_op_char": false,
    "is_word": false,
    "starti": 37,
    "value": ".",
  },
]
`;

exports[`Test test5: update document 7`] = `17`;

exports[`Test test5: update document 8`] = `38`;

exports[`Test test5: update document 9`] = `Array []`;

exports[`Test test6: update document 1`] = `"SET ACTION TRANSFER ."`;

exports[`Test test6: update document 2`] = `
Array [
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 0,
    "value": "SET",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 4,
    "value": "ACTION",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": false,
    "is_op_char": false,
    "is_word": true,
    "starti": 11,
    "value": "TRANSFER",
  },
  Object {
    "is_eoInput": false,
    "is_eoStatement": true,
    "is_op_char": false,
    "is_word": false,
    "starti": 20,
    "value": ".",
  },
]
`;

exports[`Test test6: update document 3`] = `0`;

exports[`Test test6: update document 4`] = `21`;

exports[`Test test6: update document 5`] = `Array []`;
